File: linkedlist/leetcode_Merge_Two_Sorted_Lists.java
/*Merge_Two_Sorted_Lists

Merge two sorted linked lists and return it as a new list. The new list should
be made by splicing together the nodes of the first two lists.
*/

class leetcode_Merge_Two_Sorted_Lists {
    public static void main(String[] args) {
    }
    public static ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) {
            return l2;
        }
        ListNode prev = new ListNode(0);
        prev.next = l1;
        ListNode cur = l1;
        ListNode head = prev;
        while (l2 != null) {
            if (cur.val < l2.val) {
                if (cur.next != null) {
                    // cur ptr and prev ptr move by one step.
                    prev = cur;
                    cur = cur.next;
                } else {
                    // Simply copy l2 to cur.next and done.
                    cur.next = l2;
                    break;
                }
            } else {
                // Must be: prev.val >= l2.val >= cur.val.
                ListNode temp = l2.next;
                prev.next = l2;
                l2.next = cur;
                // Move by one step.
                prev = l2;
                l2 = temp;
            }
        }
        return head.next;
    }
}

/* Python Version
def mergeTwoLists(self, a, b):
    if not a:
        return b
    if not b:
        return a

    # use a as base.
    dummy = ListNode(-1)
    dummy.next = a
    prev = dummy
    while a and b:
        if b.val >= a.val:
            a = a.next
            prev = prev.next
        else:
            nex = b.next
            prev.next = b
            b.next = a
            b = nex
            prev = prev.next

    if b:
        prev.next = b

    return dummy.next
*/

File: linkedlist/leetcode_Merge_Two_Sorted_Lists.cpp
/*
 * Copyright 2015
 * Author: madarfacar
 *
 * leetcode merge two sorted lists
 */

#include <stdio.h>
#include <memory>

using std::auto_ptr;


struct ListNode {
    int val;
    ListNode *next;
    explicit ListNode(int x) : val(x), next(NULL) {}
};


void printLinkedList(ListNode* head) {
  while (head != NULL) {
    printf("%d -> ", head->val);
    head = head->next;
  }
  printf("NULL\n");
}


class Solution {
 public:
  ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
    if (l1 == NULL) {
      return l2;
    }

    if (l2 == NULL) {
      return l1;
    }

    ListNode dummy(0);
    ListNode* cursor = &dummy;

    while (l1 != NULL && l2 !=NULL) {
      if (l1->val > l2->val) {
        cursor->next = l2;
        l2 = l2->next;
      } else {
        cursor->next = l1;
        l1 = l1->next;
      }
      cursor = cursor->next;
    }

    if (l1 != NULL) {
      cursor->next = l1;
    } else {
      cursor->next = l2;
    }

    return dummy.next;
  }
};



int main() {
  Solution sol;
  ListNode* l1 = new ListNode(1);
  ListNode* next1 = new ListNode(3);
  ListNode* nnext1 = new ListNode(5);
  l1->next = next1;
  next1->next = nnext1;

  ListNode* l2 = new ListNode(2);
  ListNode* next2 = new ListNode(4);
  ListNode* nnext2 = new ListNode(6);
  l2->next = next2;
  next2->next = nnext2;

  printf("Before:\n");
  printLinkedList(l1);
  printLinkedList(l2);

  printf("After:\n");
  ListNode* res = sol.mergeTwoLists(l1, l2);
  printLinkedList(res);
}
